controllers/flinkcluster_observer.go

```diff
@@ -163,90 +165,77 @@ func (observer *_ClusterStateObserver) observeJob(
 		observedState.job = observedJob
 	}
 
-	// Job pod.
-	var observedJobPods = new(corev1.PodList)
-	err = observer.observeJobPods(observedJobPods)
-	if err != nil {
-		if client.IgnoreNotFound(err) != nil {
-			log.Error(err, "Failed to get job pods")
-			return err
-		}
-		log.Info("Observed job pods", "pods", "nil")
-		observedJobPods = nil
-	} else {
-		log.Info("Observed job pods", "pods", observedJobPods.Items)
-		if len(observedJobPods.Items) == 1 {
-			observedState.jobPod = &observedJobPods.Items[0]
-		} else if len(observedJobPods.Items) > 1 {
-			// Should never be true unless the user manually created a pod
-			// which matches the labels.
-			return fmt.Errorf(
-				"Exactly one job pod is expected, but found more: %p",
-				observedJobPods.Items)
-		}
-	}
-
-	// Flink job ID.
-	var observedJobStatus = observedState.cluster.Status.Components.Job
-	if observedJobStatus != nil && len(observedJobStatus.ID) > 0 {
-		log.Info("Flink job ID is already available.", "ID", observedJobStatus.ID)
-		observedState.flinkJobID = &observedJobStatus.ID
-	} else {
-		var isJobCreated = observedJob != nil &&
-			observedState.jobPod != nil &&
-			observedState.jobPod.Status.Phase != corev1.PodPhase("Pending") &&
-			observedState.jobPod.Status.Phase != corev1.PodPhase("Unknown")
-		if isJobCreated && observedState.jmService != nil {
-			var url = fmt.Sprintf(
-				"http://%s.%s.svc.cluster.local:%d/jobs",
-				observedState.jmService.GetName(),
-				observedState.jmService.GetNamespace(),
-				*observedState.cluster.Spec.JobManagerSpec.Ports.UI)
-			log.Info(
-				"Polling job status from Flink API...",
-				"url",
-				url,
-				"jobPodPhase",
-				observedState.jobPod.Status.Phase)
-			var flinkJobID = observer.getFlinkJobID(url)
-			if flinkJobID != nil {
-				observedState.flinkJobID = flinkJobID
-			}
-		} else {
-			log.Info("Skip getting Flink job ID")
-		}
-	}
-
 	return nil
 }
 
-// Gets Flink job ID through Flink REST API.
-func (observer *_ClusterStateObserver) getFlinkJobID(url string) *string {
+// Observes Flink jobs through Flink API (instead of Kubernetes jobs through
+// Kubernetes API).
+//
+// This needs to be done after the cluster is running and before the job is
+// submitted, because we use it to detect whether the Flink API server is up
+// and running.
+func (observer *_ClusterStateObserver) observeFlinkJobs(
```

Comment:
Not sure if observe[foo] is a converntion is kebernetes API, but this method does bunch of things like updating job lists and job ID, the method name itself doesn't appear to match with what method itself is doing.
Maybe consider breaking method down further or replace with a more specific method name?

Reply:
"updating job lists and job ID" is applied to an intermediate object, which is used to just hold the observed state, not applied to the cluster CR.